README.md
# webmencoder

## Introduction

`webmencoder` is an opiniated `ffmpeg` wrapper tool that converts MPEG files to WebM format.

## Reason For Existence

FFmpeg is an incredibly useful tool for converting source media into other formats that are more space-efficient. As well as there being a large number of choices to make regarding the output media format, target quality/bit-rates, etc. target-device compatibility can further constrain what choices can be made. The end result of this is that the FFmpeg command-line can become quite a complex
beast, and having to construct that on a source by source basis can be both frustrating and error-prone.

This tool automates the process of invoking FFmpeg with the correct command-line options based on the input media and the opinions described below.

## Webmencoder's Opinions

### Media File Format

* Container: WebM - it's open and royalty-free. As such, it's supported by lots of open source media players and web browsers

### Video Encoding

* Codec: VP9 - it's open and royalty-free. It also has good support in open source media players and web browsers. Whilst AV1 is [recommended](https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Video_codecs#Recommendations_for_high-quality_video_presentation), it lacks [hardware encoding support](https://en.wikipedia.org/wiki/Intel_Quick_Sync_Video), at least on recent Intel CPUs, and
software encoding is still pretty slow in comparison to VP9.
* Encoding method: Hardware - best current practice suggests the use of 2-pass encoding to obtain reasonable video quality using software encoding. On my hardware, it takes on average 25% of the video duration to complete the 1st pass, and 250% to complete the 2nd pass, for a total duration of 275%. In contrast, hardware encoding completes in 16% of the video duration and requires only a single
pass to achieve similar quality levels. Using software, it will take approximately 124 minutes to encode a 45 minute video compared to just 7 minutes using hardware.
* Resolution, frame-rate, etc.: Original media resolutions will be used for the encoded video. This leaves the client device in control of any up- or down-scaling required.

### Audio Encoding

* Codec: Vorbis - Although the WebM container format supports both Vorbis and Opus codecs, some of my existing devices [don't support Opus natively](https://developer.samsung.com/smarttv/develop/specifications/media-specifications/2018-tv-video-specifications.html), resulting in my NAS having to transcode media on the fly. If it wasn't for that issue, I'd probably have selected Opus instead,
using the Fullband profile.
* Channels, sample rate, etc.: All audio streams will be included in the encoded media, using the same channel layout (7, 5.1, stereo, etc.) and sample rate as the original. The only exception to this is when the original media contains 7.1 channel DTS streams (e.g. from BluRay discs). As some of my existing devices [only support 5.1 channels in Vorbis streams](https://developer.samsung.com/smarttv/develop/specifications/media-specifications/2018-tv-video-specifications.html)
these have to be converted to 5.1 channel streams, otherwise playback is impossible to watch due to its constant buffering and stuttering. If my devices _did_ support Opus, then this utility would also change `5.1(side)` streams (commonly found on DVD media) to `5.1` due to an [FFmpeg bug](https://trac.ffmpeg.org/ticket/5718).
* Metadata: Language metadata will be added to each audio stream; this helps media players automatically select the correct stream based on a user's language preference settings.

### Subtitles

* All subtitles are currently skipped/stripped. Whilst retaining them would be in-keeping with `webmencoder`s general principle of preserving as much of the original media's contents and settings as possible, I'm yet to find a suitable open source tool that can extract subtitles from an MPEG file, OCR them, and convert them to `WebVTT` format for inclusion into the WebM container. If anyone knows
of such a tool, I'll be happy to add subtitle support!

## Usage

```bash
USAGE:
    webmencoder [OPTIONS] --input-dir <input-dir> --output-dir <output-dir>

FLAGS:
        --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
    -h, --hw-device <hw-device>       [default: /dev/dri/renderD128]
    -i, --input-dir <input-dir>
    -o, --output-dir <output-dir>
    -q, --quality <quality>           [default: 100]
```

As you can see, running `webmencoder` is rather trivial. You simply give it an input directory and an output directory and it will recursively process all `.mpeg` files it finds and convert them to `.webm` files. You can optionally provide it with a quality setting, if you find the default to be either too poor in quality or results in files larger than you'd prefer. It's also possible to tell it
what hardware device to use to perform the encoding in case the default isn't suitable.

## TODO

* Error handling
* Allow audio stream ID -> language mappings to be passed in at runtime instead of being hardcoded; these are unlikely to be consistent across input files

## License

Licensed under either of

* Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or <http://www.apache.org/licenses/LICENSE-2.0>)
* MIT license ([LICENSE-MIT](LICENSE-MIT) or <http://opensource.org/licenses/MIT>)

at your option.

## Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
# ChangeLog

## [0.1.0](https://github.com/mattburgess/webmencoder/compare/v0.0.0...v0.1.0) (2021-11-27)


### Features

* Add initial implementation and docs ([9e7d32b](https://github.com/mattburgess/webmencoder/commit/9e7d32b66c5de4422808870d012d94abc2e0636c))


### Bug Fixes

* **build:** Bump node version for release workflow ([#10](https://github.com/mattburgess/webmencoder/issues/10)) ([c083861](https://github.com/mattburgess/webmencoder/commit/c083861853c438910d377e0ce4cde9a8aa0e9862))
* **build:** Only bump our version number ([#9](https://github.com/mattburgess/webmencoder/issues/9)) ([0049021](https://github.com/mattburgess/webmencoder/commit/00490218c116029a4007c0760f74e5b601e35827))
* Handle DVDs with data streams ([96d5877](https://github.com/mattburgess/webmencoder/commit/96d5877934d252251e8eb0edb6a1abd732e34e9e))

### [0.0.2](https://github.com/mattburgess/webmencoder/compare/v0.0.1...v0.0.2) (2020-12-20)


### Bug Fixes

* **build:** Ignore ChangeLog linting errors ([09cdfda](https://github.com/mattburgess/webmencoder/commit/09cdfdadd00b304c71fda945a894ac8dd235fb35))

### [0.0.1](https://github.com/mattburgess/webmencoder/compare/v0.0.0...v0.0.1) (2020-11-16)


### Bug Fixes

* **build:** Add ChangeLog and Cargo.toml to releases ([29f0e47](https://github.com/mattburgess/webmencoder/commit/29f0e47ab3f131941b0598e788fca84e0c125071))
* **build:** Add ChangeLog title ([b864cef](https://github.com/mattburgess/webmencoder/commit/b864cefb6e071faad4a76bcd34938d274bc6dfd4))
* **build:** Fix ChangeLog linting errors ([e4aa3f5](https://github.com/mattburgess/webmencoder/commit/e4aa3f5b20163b29c82d077b17414ca3b0ee0a39))
* **build:** Fix template var interpolation ([3fcd851](https://github.com/mattburgess/webmencoder/commit/3fcd851197cad73cc72275e41095f20e99f6b534))
* **build:** Install git plugin for semantic-release ([288e7b6](https://github.com/mattburgess/webmencoder/commit/288e7b6b3bd20520a7c11f8777fea4d7f6b3c3e2))
* **build:** Publish GitHub release ([739e280](https://github.com/mattburgess/webmencoder/commit/739e280fa5a573ea9e0f4bd02cc736c462a7e13c))


src/main.rs
// SPDX-License-Identifier: (MIT OR Apache-2.0)

use glob::glob;
use serde::Deserialize;
use std::process::Command;
use structopt::StructOpt;

#[derive(StructOpt)]
#[structopt(
    name = "webmencoder",
    about = "ffmpeg wrapper to convert MPEG videos to WebM"
)]
struct Cli {
    #[structopt(short, long, parse(from_os_str))]
    input_dir: std::path::PathBuf,
    #[structopt(short, long, parse(from_os_str))]
    output_dir: std::path::PathBuf,
    #[structopt(short, long, default_value = "/dev/dri/renderD128")]
    hw_device: String,
    #[structopt(short, long, default_value = "100")]
    quality: u8,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "lowercase", tag = "codec_type")]
enum Stream {
    Video {
        field_order: String,
    },
    Audio {
        #[serde(default = "default_channel_layout")]
        channel_layout: String,
        channels: u8,
        id: String,
    },
    Subtitle,
    Data,
}

#[derive(Debug, Deserialize)]
struct Metadata {
    streams: Vec<Stream>,
}

fn default_channel_layout() -> String {
    "stereo".to_string()
}

fn get_lang_from_id(id: &str) -> &str {
    match id {
        "0x80" => "eng",
        "0x81" => "fra",
        "0x82" => "spa",
        _ => "",
    }
}

fn build_command(
    metadata: &Metadata,
    hw_device: &str,
    quality: u8,
    input_vid: &std::path::Path,
    output_dir: &std::path::Path,
) -> Result<(), std::io::Error> {
    let mut a_filters = Vec::new();
    let mut mappings = Vec::new();
    let mut a_metadata = Vec::new();
    let (mut v_idx, mut a_idx) = (0, 0);
    let mut v_filters = vec!["format=nv12", "hwupload"];
    for stream in metadata.streams.iter() {
        match stream {
            Stream::Audio {
                channel_layout,
                channels,
                id,
            } => {
                if *channels == 0 {
                    mappings.push("-map".to_string());
                    mappings.push(format!("-0:a:{}", a_idx));
                } else if channel_layout == "7.1" {
                    {
                        a_filters.push(format!(
                            "[0:a:{}]aformat=channel_layouts='5.1'['audio{}']",
                            a_idx, a_idx
                        ));
                        mappings.push("-map".to_string());
                        mappings.push(format!("[audio{}]", a_idx));
                    }
                } else {
                    mappings.push("-map".to_string());
                    mappings.push(format!("0:a:{}", a_idx));
                }
                a_metadata.push(format!("-metadata:s:a:{}", a_idx));
                a_metadata.push(format!("language={}", get_lang_from_id(id)));
                a_idx += 1;
            }

            Stream::Video { field_order } => {
                if ["tt".to_string(), "tb".to_string(), "bt".to_string()].contains(field_order) {
                    v_filters.insert(0, "yadif");
                }
                mappings.push("-map".to_string());
                mappings.push(format!("0:v:{}", v_idx));
                v_idx += 1;
            }

            Stream::Subtitle => {}

            Stream::Data => {}
        }
    }

    let in_file = input_vid.to_str().unwrap();
    let mut out_path = output_dir.join(input_vid);
    out_path.set_extension("webm");
    let out_file = out_path.to_str().unwrap();
    let a_filter = a_filters.join(";");
    let v_filter = v_filters.join(",");
    let mut ffmpeg = Command::new("ffmpeg");
    let mut ffmpeg_args = vec!["-vaapi_device", hw_device, "-i", in_file];

    if !&a_filter.is_empty() {
        ffmpeg_args.extend_from_slice(&["-filter_complex", &a_filter])
    }

    let mapping_args: Vec<&str> = mappings.iter().map(AsRef::as_ref).collect();

    ffmpeg_args.extend_from_slice(&mapping_args);

    let metadata_args: Vec<&str> = a_metadata.iter().map(AsRef::as_ref).collect();

    ffmpeg_args.extend_from_slice(&metadata_args);

    let qual = &quality.to_string();
    ffmpeg_args.extend_from_slice(&[
        "-vf",
        &v_filter,
        "-c:v",
        "vp9_vaapi",
        "-global_quality",
        qual,
        "-c:a",
        "libvorbis",
        "-y",
        out_file,
    ]);
    ffmpeg.args(ffmpeg_args);
    println!("{:?}", ffmpeg);
    ffmpeg.status().expect("Failed to convert video");
    Ok(())
}

fn get_video_details(input_vid: &std::path::Path) -> Result<Metadata, std::io::Error> {
    let ffprobe = Command::new("ffprobe")
        .args(&[
            "-v",
            "quiet",
            "-of",
            "json",
            "-show_streams",
            input_vid.to_str().unwrap(),
        ])
        .output()
        .expect("Failed to analyze input file");
    let parsed_output: Metadata = serde_json::from_slice(&ffprobe.stdout)?;
    Ok(parsed_output)
}

fn convert_file(
    input_vid: &std::path::Path,
    output_dir: &std::path::Path,
    hw_device: &str,
    quality: u8,
) -> Result<(), std::io::Error> {
    let vid_details = get_video_details(input_vid).unwrap();
    build_command(&vid_details, hw_device, quality, input_vid, output_dir)
}

fn convert_files(
    input_dir: &std::path::Path,
    output_dir: &std::path::Path,
    hw_device: &str,
    quality: u8,
) -> Result<(), std::io::Error> {
    let pattern = format!("{}{}", input_dir.to_str().unwrap(), "/**/*.mpeg");
    for entry in glob(&pattern).unwrap() {
        match entry {
            Ok(path) => convert_file(&path, output_dir, hw_device, quality).unwrap(),
            Err(e) => return Err(e.into_error()),
        }
    }
    Ok(())
}
fn main() {
    let args = Cli::from_args();
    convert_files(
        &args.input_dir,
        &args.output_dir,
        &args.hw_device,
        args.quality,
    )
    .unwrap();
}
